### DISCUSSION TITLE
Goodwe hybrid inverter setup

### DISCUSSION THREAD

**Discussion by Kveri** [slipx06/sunsynk-power-flow-card#173] [Goodwe]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

I'm trying to setup this great looking card for hybrid goodwe inverter. I have grafana running and collector collecting registers directly from the inverter, so I'm able to compare values and see if they are right. I have load connected at both 'backup' side of the inverter as well as between inverter and CTs (on-grid side).

Unfortunately I'm not able to fix three things:
1. Non-essential (On-Grid in my setup) consumption calculation: based on the docs this is calculated as 172 - 169, I believe I have 172 handled properly (1294W), so the issue seems to be in 169. I have not seen any examples on which sensor is the right one to use for 169. If I put 'zero' to 169, which is obviously wrong, then I get Non-essential power equal to Grid power. I also tried total_power and active_power, but those are not right either.
2. Blue rectangle power in the middle - I believe this should denote what power is entering/exiting the inverter from the grid side, which means that it should equal to Grid power - Non-essential power. In docs this is 169, Also that would match the 172 - 169 formula. So if I can fix 169, then both points 1 and 2 should be fixed.
3. White rectangle above the inverter - in the doc I can see that this is 175 and it is set to 'zero'. But why is that? Since Essential power without using aux is calculated as 175 + 169, which should mean: whatever power inverter gives [PV+battery] + power coming from grid to inverter, which seems right, can we calculate inverter power 175 from this formula: essential_power - 169?

Or am I missing something? What registers should be used for 169 and 175 ideally?

![image](https://github.com/slipx06/sunsynk-power-flow-card/assets/3380181/b1377bae-90e2-40e5-8b7e-a869eaa4e5de)

Thanks.

### Steps To Reproduce

N/A

### Expected behaviour

The proper values are calculated.

### Card Version

newest

### Home Assistant Version

2023.11.3

### Configuration

```shell
type: custom:sunsynk-power-flow-card
cardstyle: full
panel_mode: true
large_font: true
show_solar: true
inverter:
  modern: true
  colour: white
  autarky: energy
  model: goodwe_gridmode
  three_phase: true
  auto_scale: true
battery:
  energy: 16800
  shutdown_soc: 40
  shutdown_soc_offgrid: 25
  invert_power: false
  show_daily: true
  max_power: 10000
  full_capacity: 100
  empty_capacity: 0
  show_absolute: true
  auto_scale: true
solar:
  show_daily: true
  mppts: 2
  display_mode: 3
  max_power: 13000
  pv1_name: left PV1
  pv2_name: right PV2
  auto_scale: true
load:
  show_daily: true
  max_power: 15000
  essential_name: Backup
  additional_loads: 0
  auto_scale: true
grid:
  show_daily_buy: true
  show_daily_sell: true
  show_nonessential: true
  nonessential_name: On-Grid
  invert_grid: true
entities:
  day_battery_discharge_71: sensor.today_battery_discharge
  day_battery_charge_70: sensor.today_battery_charge
  day_load_energy_84: sensor.today_load
  day_pv_energy_108: sensor.today_s_pv_generation
  inverter_voltage_154: sensor.on_grid_l1_voltage
  inverter_voltage_L2: sensor.on_grid_l2_voltage
  inverter_voltage_L3: sensor.on_grid_l3_voltage
  load_frequency_192: sensor.meter_frequency
  inverter_current_164: sensor.on_grid_l1_current
  inverter_current_L2: sensor.on_grid_l2_current
  inverter_current_L3: sensor.on_grid_l3_current
  inverter_power_175: sensor.total_power
  pv1_power_186: sensor.pv1_power
  pv2_power_187: sensor.pv2_power
  pv_total: sensor.pv_power
  battery_voltage_183: sensor.battery_voltage
  battery_soc_184: sensor.battery_state_of_charge
  battery_power_190: sensor.battery_power
  battery_current_191: sensor.battery_current
  essential_power: sensor.back_up_load
  load_power_L1: sensor.back_up_l1_power
  load_power_L2: sensor.back_up_l2_power
  load_power_L3: sensor.back_up_l3_power
  nonessential_power: none
  non_essential_load1: sensor.on_grid_l1_power
  non_essential_load2: sensor.on_grid_l2_power
  non_essential_load3: sensor.on_grid_l3_power
  grid_power_169: sensor.total_power
  grid_ct_power_172: sensor.active_power_l1
  grid_ct_power_L2: sensor.active_power_l2
  grid_ct_power_L3: sensor.active_power_l3
  grid_ct_power_total: sensor.active_power
  pv1_voltage_109: sensor.pv1_voltage
  pv1_current_110: sensor.pv1_current
  pv2_voltage_111: sensor.pv2_voltage
  pv2_current_112: sensor.pv2_current
  grid_connected_status_194: sensor.grid_mode_code
  inverter_status_59: sensor.on_grid_mode_code
  battery_status: sensor.battery_mode_code
  total_pv_generation: sensor.total_pv_generation
  battery_temp_182: sensor.battery_temperature
  radiator_temp_91: sensor.inverter_temperature_radiator
  dc_transformer_temp_90: sensor.inverter_temperature_air
  day_grid_import_76: sensor.energy_buy_daily
  day_grid_export_77: sensor.energy_sell_daily
  remaining_solar: sensor.energy_production_today_total
  energy_cost_buy: sensor.spot_price_buy
  energy_cost_sell: sensor.spot_price_sell
```

**Comment by slipx06** [comment]
Hi. Those measurement points are specific to the Sunsynk/Deye inverters so we need to find best fit for the others. I would recommend that you specify your own sensor for non_essential power and not rely on the internal card logic i.e.`nonessential = grid_ct_power_172 + grid_ct_power_L2 + grid_ct_power_L3 - grid_power_169`

Replace `nonessential_power: none` with your `nonessential_power: sesnor.your.sesnor.that.measures.non.essential` 

The other two sensors you would need to find a best fit to represent

175 = Inverter power
169 = You can create you own template sensor for this

You can also try setting to `none` or delete to hide.
